lesson2/task1. Первая задача

Gukavina opened the pull request.

Gukavina committed:
Cделала задания из lesson1/task1.

AnastasiyaTarasova committed:
Решила первую задачу из lesson2/task2

AnastasiyaTarasova committed:
Решила первую задачу из lesson2/task2

ice-phoenix commented:
Просьба проверить соответствие имени в репозитории и имени на GitHub.

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +102,12 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
  * Например, 100 рублей под 10% годовых превратятся в 133.1 рубля
  */
-fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
+fun accountInThreeYears(initial: Int, percent: Int) = initial*(percent/100.0+1)*(percent/100.0+1)*(percent/100.0+1)
```
Общее выражение лучше вынести в отдельную переменную

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +102,12 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
  *Необходимо вывести число, полученное из заданного перестановкой цифр в обратном порядке (например, 874).
  */
-fun numberRevert(number: Int): Int = TODO()
+fun numberRevert(number: Int) = number%10*100+number/10%10*10+number/100%10
```
Отдельные подвыражения лучше вынести в переменные с осмысленными именами

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +33,32 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
  */
-fun ageDescription(age: Int): String = TODO()
+fun ageDescription(age: Int): String {
+    if (age % 10 == 1) {
```
Подумайте над тем, как записать данную цепочку `if`'ов в виде `when`

Gukavina committed:
Сделала полностью lesson2/task1

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +69,26 @@ fun ageDescription(age: Int): String = TODO()
+    if (sAll<t1*v1){
+        val tHalf = sAll/v1
+        return(tHalf)
+    } else {
```
`else { if ... }` лучше оформлять как `else if ...`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +100,17 @@ fun timeForHalfWay(t1: Double, v1: Double,
                        rookX1: Int, rookY1: Int,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+    if ((rookX1 == kingX)or(rookY1 == kingY)or(rookX2 == kingX)or(rookY2 == kingY)){
```
Факт того, что какая-то фигура бьет короля, лучше вынести в отдельные переменные

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +123,18 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
                           rookX: Int, rookY: Int,
-                          bishopX: Int, bishopY: Int): Int = TODO()
+                          bishopX: Int, bishopY: Int): Int {
+    if(((kingX-kingY) == (bishopX-bishopY))or(bishopX-kingX == bishopY-kingY)or(bishopX-kingX == -(bishopY-kingY))){
```
Факт того, что какая-то фигура бьет короля, лучше вынести в отдельные переменные

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +144,21 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
  */
-fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+fun triangleKind(a: Double, b: Double, c: Double): Int {
+    if ((a+b>c)and(b+c>a)and(a+c>b)){
```
Удобнее сперва найти максимальную сторону треугольника, а потом все считать относительно нее

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +168,18 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
-fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    if (a > d) return (-1)
```
Данное выражение лучше оформить как `when`

ice-phoenix commented:
Разберитесь с тем, что ваш код для `timeForHalfWay` должен возвращать в случае, если итоговая пройденная дистанция равна 0.

Gukavina committed:
Сделала все, кроме одной, задачи из lesson2/task2

ice-phoenix commented:
Просьба исправить **все** приведенные к задачам замечания перед подачей следующих PR

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +17,27 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
+fun isNumberHappy(number: Int): Boolean {
+    val sum34 = number%10+(number/10)%10
+    val sum12 = number/100%10+number/1000
+    if (sum34 == sum12) {
```
Выражение `if (cond) true else false` можно заменить на `cond`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +17,27 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
  */
-fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
+fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean {
+    if ((x1 == x2) or (y1 == y2) or (x1-x2 == y1-y2) or (x1-x2 == y2-y1))
```
- Выражение `(x1-x2 == y1-y2) or (x1-x2 == y2-y1)` можно посчитать проще, если вспомнить о такой функции, как модуль числа
- Выражение `if (cond) true else false` можно заменить на `cond`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +58,11 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
+    val sAB = a*b
+    val sCB = c*b
+    val sAC = a*c
+    if ((sAB<=sHole)or(sCB<=sHole)or(sAC<=sHole)) return true
```
Выражение `if (cond) true else false` можно заменить на `cond`

Gukavina committed:
Переделала brickPasses из lesson2/task2

Gukavina committed:
В Lesson2/Task1 переделала ageDescription, используя конструкцию "when". В timeForHalfWay добавила условие, когда пройденный путь равен нулю, выводит tHalf=0.0

Gukavina committed:
В Lesson2/Task2 переделала brickPasses и сделала circleInside. Закончила второй урок

Gukavina committed:
В Lesson2/Task1  исправила ageDescription в файле тестов добавила проверку на "2", которую не прошла программа после PR

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +34,15 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+    return when {
+        (age % 100 == 11) -> "$age лет"
+        (age % 10 == 1)and(age % 100 !== 11) -> "$age год"
+        ((age % 100 < 15) and (age % 100 > 11)) -> "$age лет"
```
Подумайте, нельзя ли упростить некоторые из данных условий, например, объединив их вместе?

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +53,31 @@ fun ageDescription(age: Int): String = TODO()
                    t2: Double, v2: Double,
-                   t3: Double, v3: Double): Double = TODO()
+                   t3: Double, v3: Double): Double  {
+    val sAll = (t1*v1+t2*v2+t3*v3)/2
```
Повторно используемые выражения лучше вынести в отдельные переменные

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +112,18 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
                           rookX: Int, rookY: Int,
-                          bishopX: Int, bishopY: Int): Int = TODO()
+                          bishopX: Int, bishopY: Int): Int {
+    if(((kingX-kingY) == (bishopX-bishopY))or((bishopX-kingX) == (bishopY-kingY))or(bishopX-kingX == -(bishopY-kingY))){
```
Факт того, что какая-то фигура бьет короля, лучше вынести в отдельные переменные

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -35,7 +47,11 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
 fun circleInside(x1: Double, y1: Double, r1: Double,
-                 x2: Double, y2: Double, r2: Double): Boolean = TODO()
+                 x2: Double, y2: Double, r2: Double): Boolean {
+    if ((Math.sqrt(sqr(x2 - x1) + sqr(y2 - y1)) + r1) <= r2) return true
```
Выражение `if (cond) true else false` можно заменить на `cond`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +62,8 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
-fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = TODO()
+fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean {
+
+    if ((((s>=a)and(r>=c))or((r>=a)and(s>=c)))or(((s>=b)and(r>=a))or((r>=b)and(s>=a)))or(((s>=b)and(r>=c))or((r>=b)and(s>=c)))) return true
```
Выражение `if (cond) true else false` можно заменить на `cond`

ice-phoenix replied:
Лучше сперва найти две минимальные стороны кирпича, а потом все считать относительно них

ice-phoenix reviewed (changes requested):
Просьба исправлять замечания перед подачей следующих PR

Gukavina committed:
В Lesson2/Task1:
1. В ageDescription упростила условия.
2. В timeForHalfWay вынесла вырожения, которые используются несколько раз в отдельные переменные.
3. В rookOrBishopThreatens и whichRookThreaten вынесла факт того, что фигура бъёт короля в отдельные переменные типа Boolean
В Lesson2/Task2:
В circleInside и brickPasses убрала ветвления.

Gukavina committed:
Сделала тривиальные, простые и среднии задачи из Lesson3

ice-phoenix commented:
Ваши PR не будут проверяться до тех пор, пока вы не исправите замечания к предыдущим PR.

Gukavina committed:
Переделала ошибочные задачи из Lesson3

Gukavina committed:
Марат Халимович, я уже исправила ошибки,  на которые Вы указали в письме 29 сентября 11:52,отправила PR  сегодня утром на паре. По каждому пункту сделала исправление.
Дублирую комит:
В Lesson2/Task1:
1. В ageDescription упростила условия.
2. В timeForHalfWay вынесла вырожения, которые используются несколько раз в отдельные переменные.
3. В rookOrBishopThreatens и whichRookThreaten вынесла факт того, что фигура бъёт короля в отдельные переменные типа Boolean
В Lesson2/Task2:
В circleInside и brickPasses убрала ветвления.

В Lesson3 сделала одну сложную, с sin и cos пока не разобралась.

Gukavina committed:
В Lesson3 разобралась с cos и sin

ice-phoenix commented:
Если мы с вами посмотрим во вкладку `Files changed`, то увидим там достаточно большое количество комментариев, которые вы пропустили при исправлении. Убедительная просьба
- пользоваться страницей `Files changed` для просмотра замечаний преподавателя
- использовать систему комментариев к PR для общения с преподавателем, а не писать вопросы в коммит-сообщение (задача коммит-сообщения --- кратко описывать то, что было сделано в данном коммите)
- исправить те замечания, которые были вами пропущены

Gukavina committed:
Мне кажется, что я исправила имеющиеся замечания в lesson1 и Lesson2; что ещё мне нужно исправить?

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +35,18 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+    val ten = age % 10
+    val oneHundred = age % 100
+    return when {
+        (ten == 1) -> when { (oneHundred == 11) -> "$age лет"
```
Вложенный `when` лучше унести на новую строку

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +57,34 @@ fun ageDescription(age: Int): String = TODO()
+    val s1 = t1 * v1
+    val s2 = t2 * v2
+    val s12 = s1 + s2
+    if (sAll > 0) {
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +96,19 @@ fun timeForHalfWay(t1: Double, v1: Double,
                        rookX1: Int, rookY1: Int,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+    val rook1Danger: Boolean = (rookX1 == kingX) or (rookY1 == kingY)
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +96,19 @@ fun timeForHalfWay(t1: Double, v1: Double,
+                       rookX2: Int, rookY2: Int): Int {
+    val rook1Danger: Boolean = (rookX1 == kingX) or (rookY1 == kingY)
+    val rook2Danger: Boolean = (rookX2 == kingX) or (rookY2 == kingY)
+    if ((rook1Danger == true) or (rook2Danger == true)) {
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +121,20 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
                           rookX: Int, rookY: Int,
-                          bishopX: Int, bishopY: Int): Int = TODO()
+                          bishopX: Int, bishopY: Int): Int {
+    val bishopDanger: Boolean = ((kingX - kingY) == (bishopX - bishopY)) or ((bishopX - kingX) == (bishopY - kingY)) or (bishopX - kingX == -(bishopY - kingY))
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +121,20 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+                          bishopX: Int, bishopY: Int): Int {
+    val bishopDanger: Boolean = ((kingX - kingY) == (bishopX - bishopY)) or ((bishopX - kingX) == (bishopY - kingY)) or (bishopX - kingX == -(bishopY - kingY))
+    val rookDanger: Boolean = (rookX == kingX) or (rookY == kingY)
+    if (bishopDanger == true) {
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +144,39 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+    var max = 0.0
+    var min1 = 0.0
+    var min2 = 0.0
+    if (a >= b) {
```
Все это лучше сделать как `if (a >= b && a >= c) ... else if (b >= a && b >= c) ... else ...`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +186,19 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+            else -> 0
+        }
+    }
+    return (result)
```
Можно сразу `return when ...`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +186,19 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
-fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    val result = when {
```
Все вложенные `when` можно записать условиями `when` на первом уровне

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +18,21 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
+fun isNumberHappy(number: Int): Boolean {
+    val sum34 = number % 10 + (number / 10) % 10
+    val sum12 = number / 100 % 10 + number / 1000
+    return (sum34 == sum12)
```
Лишние скобки

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +54,4 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
  * Вернуть true, если кирпич пройдёт
  */
-fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = TODO()
+fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = ((((s >= a) and (r >= c)) or ((r >= a) and (s >= c))) or (((s >= b) and (r >= a)) or ((r >= b) and (s >= a))) or (((s >= b) and (r >= c)) or ((r >= b) and (s >= c))))
```
Данное условие лучше перенести на следующую строчку + хорошо было бы разнести отдельные подусловия на отдельные строчки

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun fib(n: Int): Int = TODO()
+fun fib(n: Int): Int {
+    val nn=n-1
```
Форматирование кода

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+            row += x
+        }
+        val result = row[nn]
+        return (result)
```
Лишние скобки

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  * минимальное число k, которое делится и на m и на n без остатка
  */
-fun lcm(m: Int, n: Int): Int = TODO()
+fun nod(x:Int, y: Int): Int = if (x!= 0) nod(y%x,x) else y
```
`nod ==> gcd` + форматирование кода

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun nod(x:Int, y: Int): Int = if (x!= 0) nod(y%x,x) else y
+fun lcm(m: Int, n: Int): Int {
+    val result = m*n/nod(m,n)
+    return(result)
```
Можно все сделать one-liner'ом через `fun foo() = ...` синтаксис

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun minDivisor(n: Int): Int {
+    var i = 2
+    while(n%i !=0) i++
+    return(i)
```
Форматирование + скобки

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun maxDivisor(n: Int): Int {
+    var i = n-1
+    while(n%i !=0) i--
+    return(i)
```
Форматирование + скобки

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +130,16 @@ fun maxDivisor(n: Int): Int = TODO()
  */
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
+    var result = 0
```
Можно ли решить эту задачу через уже реализованные вами функции?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -105,7 +148,8 @@ fun isCoPrime(m: Int, n: Int): Boolean = TODO()
  */
-fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+fun sqr(x:Int): Int = x*x
+fun squareBetweenExists(m: Int, n: Int): Boolean = ((sqr(Math.sqrt(n.toDouble()).toInt())>=m)and(sqr(Math.sqrt(n.toDouble()).toInt())<=n))
```
Форматирование + для округления лучше использовать специальные функции `round / floor / ceil`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +158,19 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
  */
-fun sin(x: Double, eps: Double): Double = TODO()
+fun sin(x: Double, eps: Double): Double {
+    var mark=1
```
Форматирование + скобки

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +179,27 @@ fun sin(x: Double, eps: Double): Double = TODO()
  * Нужную точность считать достигнутой, если очередной член ряда меньше eps по модулю
  */
-fun cos(x: Double, eps: Double): Double = TODO()
+fun cos(x: Double, eps: Double): Double = sin(x+Math.PI/2,eps)
```
Браво! Хоть кто-то вспомнил про тригонометрию!

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +179,27 @@ fun sin(x: Double, eps: Double): Double = TODO()
+        nn/=10
+        x/=10
+    }
+    return(result)
```
Форматирование + скобки

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +208,30 @@ fun revert(n: Int): Int = TODO()
+    n2=n2*10+n1%10
+    n1/=10
+}
+return (n==n2)
```
Форматирование + скобки

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +208,30 @@ fun revert(n: Int): Int = TODO()
-fun isPalindrome(n: Int): Boolean = TODO()
-
+fun isPalindrome(n: Int): Boolean {
+var n1=n
```
Можно ли данную задачу решить через уже реализованные вами ранее функции?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +240,28 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+        number += amount(nn * nn)
+    }
+    result = nn * nn
+    (n..number - 1).forEach { i -> result /= 10 }
```
Менять внешнюю переменную внутри лямбды --- **очень плохой** стиль, лучше воспользоваться старым добрым циклом `for`

ice-phoenix reviewed (changes requested):
Исправьте замечания перед продолжением работы над следующими задачами

Gukavina committed:
Исправила замечания второго и третьего урока от 6ого октября

Gukavina committed:
Исправила timeForHalfWay в уроке два.

Gukavina committed:
Четвёртый урок, кроме последних сложных

Gukavina committed:
+ 2 сложные из четвёртого

Gukavina committed:
+ 2 сложные из четвёртого(2)

Gukavina committed:
добавила условие в convert

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -93,7 +94,7 @@ fun thirdDigit(number: Int): Int = TODO()
  * Определите время поезда в пути в минутах (в данном случае 216).
  */
-fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = TODO()
+fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int) = (hoursArrive * 60 + minutesArrive) - (hoursDepart * 60 + minutesDepart)
```
Длинные строчки обычно разбивают на несколько более коротких, в данном случае, можно перенести строку после знака `=`

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +103,19 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
  */
-fun numberRevert(number: Int): Int = TODO()
+fun numberRevert(number: Int): Int {
+    val numberConversely = number % 10 * 100 + number / 10 % 10 * 10 + number / 100 % 10
```
Можно обойтись без промежуточной переменной и записать функцию как one-liner

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +35,19 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+    val ten = age % 10
+    val oneHundred = age % 100
+    return when {
+        (ten == 1) ->
```
Лучше записать плоско как

```
when {
    ten == 1 && oneHundred == 11 -> ...
    ten == 1 -> ...
    ...
}
```

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +58,23 @@ fun ageDescription(age: Int): String = TODO()
                    t2: Double, v2: Double,
-                   t3: Double, v3: Double): Double = TODO()
+                   t3: Double, v3: Double): Double {
+    val sAll = (t1 * v1 + t2 * v2 + t3 * v3) / 2
```
Лучше посчитать `sAll` из отдельных составных частей, которые вы и так считаете ниже

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +58,23 @@ fun ageDescription(age: Int): String = TODO()
+    val s2 = t2 * v2
+    val s12 = s1 + s2
+    return when {
+        (sAll > 0) ->
```
Вот эту проверку лучше, наоборот, вытащить наружу (так как она проверяет входные данные на допустимость) в виде отдельного `if` с `return`

ice-phoenix replied:
Но, в принципе, так тоже неплохо

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +86,19 @@ fun timeForHalfWay(t1: Double, v1: Double,
+                       rookX2: Int, rookY2: Int): Int {
+    val rook1Danger = (rookX1 == kingX) or (rookY1 == kingY)
+    val rook2Danger = (rookX2 == kingX) or (rookY2 == kingY)
+    return when {((rook1Danger == true) or (rook2Danger == true)) ->
```
- Весь этот `when` можно записать плоско и проще, если вспомнить, что невыполнение какой-либо ветки из `when` означает, что ее условие точно **не** выполняется
- Вместо проверки переменных типа `Boolean` на равенство `true / false` лучше просто оставить `rook1Danger / !rook1Danger`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +111,17 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+                          bishopX: Int, bishopY: Int): Int {
+    val bishopDanger = ((kingX - kingY) == (bishopX - bishopY)) or ((bishopX - kingX) == (bishopY - kingY)) or (bishopX - kingX == -(bishopY - kingY))
+    val rookDanger = (rookX == kingX) or (rookY == kingY)
+    return when {(bishopDanger == true) and (rookDanger == true) -> 3
```
См. замечание к `whichRookThreatens`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +131,35 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+        min1 = b
+        min2 = c
+
+    } else {
```
`else if (...)`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +131,35 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+    }
+    if (min1 + min2 > max) {
+        val cosMax = (min1 * min1 + min2 * min2 - max * max) / 2 * min1 * min2
+        if (cosMax >= 0.0) {
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +169,22 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    return when {
+        (a > d) or ((a < d) and (c > b)) -> -1
+        else ->
```
`else -> when ...` можно заменить, вытащив все условия вложенного `when` во внешний

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +169,22 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    return when {
+        (a > d) or ((a < d) and (c > b)) -> -1
```
[Здесь и много где] Лучше использовать `&& / ||` вместо `and / or`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +169,22 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+                    when { (c < a) -> b - a
+                        else -> b - c
+                    }
+                else -> 0
```
Если вы внимательно посмотрите на условия, то сможете увидеть, как можно упростить вычисление точек, которые участвуют в вычислении длины пересечения

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +18,21 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
  */
-fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
+fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean {
+    return ((x1 == x2) or (y1 == y2) or (Math.abs(x1 - x2) == Math.abs(y1 - y2)))
```
Функцию с телом в виде одного `return expr` можно записать как `fun foo(...) = expr`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +54,7 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
  */
-fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = TODO()
+fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean =
+        ((((s >= a) and (r >= c)) or ((r >= a) and (s >= c)))
```
Лишние скобки вокруг выражений

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,70 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun digitNumber(n: Int): Int = TODO()
+fun digitNumber(n: Int): Int {
+    var nn: Int = n
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,70 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+            row += x
+        }
+        val result = row[nn]
+        return result
```
Можно сразу `return row[nn]`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,70 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+
+fun gcd(x: Int, y: Int): Int = if (x != 0) gcd(y % x, x) else y
+
+fun lcm(m: Int, n: Int): Int = m * n / gcd(m, n)
```
:+1:

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,70 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun maxDivisor(n: Int): Int = TODO()
+fun maxDivisor(n: Int): Int {
+    var i = n - 1
```
Нет ли более точной верхней границы для возможных делителей числа `n`, чем `n - 1`?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +130,21 @@ fun maxDivisor(n: Int): Int = TODO()
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
+    val gcdMN = gcd(m, n)
+    if (gcdMN == 1) return true
```
Выражение `if (cond) return true else return false` можно заменить на `return cond`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +130,21 @@ fun maxDivisor(n: Int): Int = TODO()
+    if (gcdMN == 1) return true
+    else return false
+}
+//  var result = 0
```
Устаревшие комментарии лучше удалять

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -105,7 +153,9 @@ fun isCoPrime(m: Int, n: Int): Boolean = TODO()
-fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+fun sqr(x: Double): Double = x * x
+
+fun squareBetweenExists(m: Int, n: Int): Boolean = sqr(Math.floor(Math.sqrt(n.toDouble()))) >= m && sqr(Math.floor(Math.sqrt(n.toDouble()))) <= n
```
Повторно вычисляемое выражение лучше вынести в отдельную переменную

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +164,19 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+        sin += mark * newElement
+        ind += 2
+        newElement = Math.pow(x % (Math.PI * 2), ind.toDouble()) / factorial(ind)
+        mark = 0 - mark
```
Лучше `mark = -mark`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +164,19 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+    while (Math.abs(newElement) >= eps) {
+        sin += mark * newElement
+        ind += 2
+        newElement = Math.pow(x % (Math.PI * 2), ind.toDouble()) / factorial(ind)
```
Зачем постоянно вычислять `x % (2 * PI)`?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +185,27 @@ fun sin(x: Double, eps: Double): Double = TODO()
  * Нужную точность считать достигнутой, если очередной член ряда меньше eps по модулю
  */
-fun cos(x: Double, eps: Double): Double = TODO()
+fun cos(x: Double, eps: Double): Double = sin(x + Math.PI / 2, eps)
```
:+1:

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +185,27 @@ fun sin(x: Double, eps: Double): Double = TODO()
+    var result = 0
+    var nn = n
+    var x = 10
+    while (nn % x != n) x *= 10
```
Может быть, это можно выразить через возведение в степень?

ice-phoenix replied:
Ап

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +214,26 @@ fun revert(n: Int): Int = TODO()
-fun isPalindrome(n: Int): Boolean = TODO()
+fun isPalindrome(n: Int): Boolean {
+    val nn = revert(n)
+    return n == nn
```
Можно сразу `... = revert(n) == n`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +242,36 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
  */
-fun squareSequenceDigit(n: Int): Int = TODO()
+fun amount(n: Int): Int {
+    var number = 0
```
Нет ли у вас уже функции с точно такой же функциональностью?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +242,36 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+    }
+     */
+    var result = nn * nn
+    (n..number - 1).forEach { i -> result /= 10 }
```
Это лучше записать через старый добрый цикл `for`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +242,36 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+            number += amount(nn * nn)
+        } else break
+    }
+    /**
```
Устаревший комментарий

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,41 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun abs(v: List<Double>): Double = TODO()
+
+fun squareList(list: List<Double>) = list.map { it * it }
```
Данная функция является излишней, намного понятнее везде, где нужен ее вызов, просто писать `list.map{ it * it }`

ice-phoenix replied:
Ап

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,41 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun mean(list: List<Double>): Double = TODO()
+fun mean(list: List<Double>): Double {
+    if (list.isNotEmpty()) {
```
Лучше записать это как

```
if (...) return list.sum() / list.size
else return 0.0
```

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +106,41 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+    if (list.isNotEmpty()) {
+        val average = list.sum() / list.size
```
Если записать `average` через `mean(...)`, то можно обойтись без проверки `list.isNotEmpty()`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -138,7 +169,15 @@ fun times(a: List<Double>, b: List<Double>): Double = TODO()
  */
-fun polynom(p: List<Double>, x: Double): Double = TODO()
+fun polynom(p: List<Double>, x: Double): Double {
+    if (p.isEmpty()) return 0.0
```
Зачем нужна эта проверка? Что вернет код ниже, если список пуст?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -148,7 +187,19 @@ fun polynom(p: List<Double>, x: Double): Double = TODO()
+        for (i in 1..list.size - 1) {
+            forElemtnt = list[i]
+            list[i] = list[i] + sum
+            sum += forElemtnt
```
Зачем вам переменная `sum`, если ее значение всегда имеет элемент `list[i - 1]`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +208,28 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
-fun factorize(n: Int): List<Int> = TODO()
+fun factorize(n: Int): List<Int> {
+    var nn = n
+    var minDivisor = 0
```
Почему эта переменная объявлена здесь, а не по месту использования?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +208,28 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
-fun factorizeToString(n: Int): String = TODO()
+fun factorizeToString(n: Int): String {
+    val result = factorize(n)
+    return result.joinToString(separator = "*")
```
Можно оформить данную функцию как one-liner

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +238,17 @@ fun factorizeToString(n: Int): String = TODO()
+            result.add(0, nn % base)
+            nn /= base
+        }
+    } else result.add(0, 0)
```
Почему не просто `add(0)`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +238,17 @@ fun factorizeToString(n: Int): String = TODO()
+    if (n != 0) {
+        var nn = n
+        while (nn > 0) {
+            result.add(0, nn % base)
```
Вместо добавления в начало списка лучше добавлять в конец, а потом развернуть итоговый список при помощи `reversed()`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +258,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    val number = convert(n, base)
+    for (i in 0..number.size - 1) {
+        if (number[i] >= 10) result += (number[i] + 87).toChar()
+        else result += "${number[i]}"
```
Лучше `toString()`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +258,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
  */
-fun convertToString(n: Int, base: Int): String = TODO()
+fun convertToString(n: Int, base: Int): String {
+    var result: String = ""
```
- Система типов Котлина выведет тип переменной автоматически
- Для сборки строки из составных частей лучше использовать список строк или `StringBuilder`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +258,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    var result: String = ""
+    val number = convert(n, base)
+    for (i in 0..number.size - 1) {
+        if (number[i] >= 10) result += (number[i] + 87).toChar()
```
Что такое 87? Почему не 42?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +276,13 @@ fun convertToString(n: Int, base: Int): String = TODO()
+fun decimal(digits: List<Int>, base: Int): Int {
+    var result = 0
+    for (i in 0..digits.size - 1) {
+        result += digits[i] * Math.pow(base.toDouble(), (digits.size - 1 - i).toDouble()).toInt()
```
Напишите функцию `pow(Int, Int): Int`, работающую в целых числах, и воспользуйтесь ей везде, где вы не совсем корректно используете работу в числах с плавающей запятой для возведения в степень

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +293,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
-fun decimalFromString(str: String, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    var list: List<Int>
+    list = listOf()
```
Почему разделено определение и объявление переменной?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +293,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+    val string = str
+    for (i in 0..string.length - 1)
+        if (string[i] in '0'..'9') list += ((string[i]).toInt() - 48)
+        else list += ((string[i]).toInt() - 87)
```
Магические константы

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +293,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    var list: List<Int>
+    list = listOf()
+    val string = str
```
Зачем вам эта переменная?

ice-phoenix reviewed (changes requested):
Довольно много замечаний, исправьте их, пожалуйста, перед продолжением работы над следующими уроками

Gukavina committed:
Исправила замечания к первому, второму и третьему урокам.  Четвёртый в процессе

ice-phoenix commented:
Ваш код для `russian()` не завершается. Просьба проверить его локально на различных входных данных.

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -1,7 +1,10 @@
 
 import lesson1.task1.discriminant
+import lesson3.task1.cos
+import lesson4.task1.abs
```
Зачем?

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +57,21 @@ fun ageDescription(age: Int): String = TODO()
+    val s2 = t2 * v2
+    val s12 = s1 + s2
+    val sAll = (s1 + s2 + t3 * v3) / 2
+    if (sAll > 0) {
```
Это условие можно сделать частью `when`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +83,19 @@ fun timeForHalfWay(t1: Double, v1: Double,
+                       rookX2: Int, rookY2: Int): Int {
+    val rook1Danger = (rookX1 == kingX) or (rookY1 == kingY)
+    val rook2Danger = (rookX2 == kingX) or (rookY2 == kingY)
+    return when {(rook1Danger || rook2Danger) ->
```
Этот `when` очень сильно переусложнен. Подумайте еще раз, какие простые логические условия соответствуют возврату из функции 1, 2 и 3.

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +108,18 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+        bishopDanger && !rookDanger -> 2
+        !bishopDanger && rookDanger -> 1
+        !bishopDanger && !rookDanger -> 0
+        else -> 0
```
Зачем?

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +129,39 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+    var max = 0.0
+    var min1 = 0.0
+    var min2 = 0.0
+    when {((a >= b) && (a >= c)) -> {
```
Форматирование кода

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +129,39 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+        }
+    }
+    val cosMax = (min1 * min1 + min2 * min2 - max * max) / 2 * min1 * min2
+    return when { (min1 + min2 > max) ->
```
Эти два `when` можно объединить в один

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +171,14 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
-fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    return when {
```
Функцию с телом в виде одного `return expr` можно записать как `fun foo(...) = expr`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun digitNumber(n: Int): Int = TODO()
+fun digitNumber(n: Int): Int {
+    var nn: Int = n
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun digitNumber(n: Int): Int = TODO()
+fun digitNumber(n: Int): Int {
+    var nn: Int = n
+    var count: Int = 0
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  * Для заданного числа n > 1 найти максимальный делитель, меньший n
  */
-fun maxDivisor(n: Int): Int = TODO()
+fun maxDivisor(n: Int): Int = n / minDivisor(n)
```
:+1:

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +131,10 @@ fun maxDivisor(n: Int): Int = TODO()
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
+    val gcdMN = gcd(m, n)
+    return (gcdMN == 1)
```
Функцию с телом в виде одного `return expr` можно записать как `fun foo(...) = expr`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -105,7 +143,12 @@ fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+
+fun squareBetweenExists(m: Int, n: Int): Boolean {
+    val sqrN = Math.floor(Math.sqrt(n.toDouble()))
+    return sqr(sqrN) >= m && sqr(sqrN) <= n
```
Зачем проверять вторую часть условия?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +234,22 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+    }
+
+    var result = nn * nn
+    (n..number - 1).forEach { i -> result /= 10 }
```
Нельзя ли заменить этот цикл на возведение в степень и деление?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -148,7 +188,19 @@ fun polynom(p: List<Double>, x: Double): Double = TODO()
+fun accumulate(list: MutableList<Double>): MutableList<Double> {
+    if (list.isEmpty()) return list
+    else {
+        var forElement: Double
```
Почему эта переменная объявлена здесь, а не в теле цикла?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +209,28 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
-fun factorize(n: Int): List<Int> = TODO()
+fun factorize(n: Int): List<Int> {
+    var nn = n
+    var minDivisor: Int
```
Почему эта переменная объявлена здесь, а не в теле цикла?

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +62,33 @@ fun main(args: Array<String>) {
+    val parts = str.split(" ")
+    var result = ""
+    var f = 0
+    for (part in parts) {
```
Если вам необходимо сделать разные вещи с разными элементами списка, зачем цикл???

Gukavina committed:
Снова исправления

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +34,17 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+    val oneHundred = age % 100
+    return when {
+        (ten == 1) && (oneHundred == 11) -> "$age лет"
+        (ten == 1) && (oneHundred != 11) -> "$age год"
```
Когда это условие может выполниться?

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +105,17 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
                           rookX: Int, rookY: Int,
-                          bishopX: Int, bishopY: Int): Int = TODO()
+                          bishopX: Int, bishopY: Int): Int {
+    val bishopDanger = ((kingX - kingY) == (bishopX - bishopY)) or ((bishopX - kingX) == (bishopY - kingY)) or (bishopX - kingX == -(bishopY - kingY))
```
`or -> ||`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +163,13 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
-fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = when {
+        (a > d) || ((a < d) && (c > b)) -> -1
```
Это условие можно упростить

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +18,19 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
  * Определить, угрожают ли они друг другу. Вернуть true, если угрожают.
  */
-fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
+fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = ((x1 == x2) or (y1 == y2) or (Math.abs(x1 - x2) == Math.abs(y1 - y2)))
```
`or -> ||`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +133,8 @@ fun maxDivisor(n: Int): Int = TODO()
  * Например, 25 и 49 взаимно простые, а 6 и 8 -- нет.
  */
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean = (gcd(m, n) == 1)
```
Лишние скобки вокруг законченных выражений

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +232,21 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+fun squareSequenceDigit(n: Int): Int {
+    var nn = 0
+    var number = 0
+    for (i in 0..n) {
```
Почему бы не оформить этот цикл как `while`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -1,7 +1,11 @@
 package lesson4.task1
 
 import lesson1.task1.discriminant
+import kotlin.comparisons.reverseOrder
```
Зачем?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -148,7 +183,14 @@ fun polynom(p: List<Double>, x: Double): Double = TODO()
  */
-fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+fun accumulate(list: MutableList<Double>): MutableList<Double> {
+    if (list.isEmpty()) return list
```
Зачем данная проверка?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +199,26 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    var nn = n
+    var result = listOf<Int>()
+    while (nn > 1) {
+        var minDivisor: Int
```
Почему вы разделили определение и объявление переменной?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +247,19 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    return result
+}
+/** 87, а не 42, потому что это не ответ на главный вопрос жизни, вселенной и всего такого, а всего лишь перевод числа 10 в символ a, с использованием таблицы ASCII
+ * И если поменять toChar на toString, то всё сломается :'(
```
Вообще-то, переменные типа `Char` тоже можно вычитать и складывать, например `'a' - 10 + i`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +268,24 @@ fun convertToString(n: Int, base: Int): String = TODO()
-fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun pow(base: Int, n: Int): Int {
+    var result = 1
+    if ((base != 0) && (n != 0)) {
```
`0^0 == 1`???

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,8 +296,14 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    var list = listOf<Int>()
+    for (i in 0..str.length - 1)
+        if (str[i] in '0'..'9') list += ((str[i]).toInt() - 48)
```
См. замечание выше про работу с `Char`

Gukavina committed:
Снова исправления по второму, третьему и четвёртому

Gukavina committed:
пятый урок, две средние задачи

Gukavina committed:
пятый урок, две средние задачи.(2)

Gukavina committed:
пятый урок, две средние задачи.(2)

Gukavina committed:
поправила

mglukhikh commented:
recheck all

Gukavina committed:
Так bestLongJump должна работать

Gukavina committed:
Так bestLongJump должна работать(2)

Gukavina committed:
забыла лишнее условие

Gukavina committed:
Начала шестой

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +54,22 @@ fun ageDescription(age: Int): String = TODO()
+    if (sAll > 0) {
+        return when {
+            (sAll <= s1) -> sAll / v1
+            (sAll > s1) and (sAll <= s12) -> (t1 + (sAll - s1) / v2)
```
* Первую часть этой проверки можно не делать, вы ее уже проверили строчкой выше
* `and` ==> `&&`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,8 +124,36 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+    }
+    val cosMax = (min1 * min1 + min2 * min2 - max * max) / 2 * min1 * min2
+    return when {
+        (min1 + min2 > max) && (cosMax > 0.0) -> 0
```
Первую часть лучше проверить один раз в самом начале `when`, чем проверять ее снова и снова

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +60,71 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun maxDivisor(n: Int): Int = n / minDivisor(n)
+/*{
+    var i = n / minDivisor(n)
+    while (n % i != 0) i--
```
Ненужный более код надо не комментировать, а удалять

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +157,20 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+    var ind = 1
+    var sin = 0.0
+    var newElement = x % (Math.PI * 2)
+    val piX = x % (Math.PI * 2)
```
Зачем два раза считать одно и то же?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +232,21 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+    var nn = 0
+    var number = 0
+    for (i in 0..n) { //В прошлых исправлениях Вы писали "Менять внешнюю переменную внутри лямбды --- очень плохой стиль, лучше воспользоваться старым добрым циклом for"
+        if (number < n) {
```
Почему бы это не сделать более подходящим циклом `while`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -1,8 +1,11 @@
 
 import lesson1.task1.discriminant
 
+//import lesson3.task1.minDivisor
```
Зачем?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +107,38 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+
+
+fun abs(v: List<Double>): Double {
+    val newList = v.map { it * it }
```
Можно обойтись без этой вспомогательной переменной, а если уж с ней, то имя у нее должно быть нормальное, а не ничего не говорящее `newList`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -128,7 +147,17 @@ fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+    if (a.isEmpty() or b.isEmpty()) {
+        return 0.0
+    } else {
+        var result = 0.0
```
Подобные конструкции (когда в ветке then вы выходите из функции из-за неправильных входных данных или прочей ошибки) обычно оформляют как
```
if (cond) return ...
var result = 0.0
...
```
Выносить вторую часть в ветку else не имеет особого смысла и только усложняет понимание кода.

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +196,25 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    var nn = n
+    var result = listOf<Int>()
+    while (nn > 1) {
+        val minDivisor = lesson3.task1.minDivisor(nn)
```
Зачем указывать путь полностью?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +243,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+fun convertToString(n: Int, base: Int): String {
+    var result: String = ""
+    val number = convert(n, base)
+    for (i in 0..number.size - 1) {
```
Если вам не нужен индекс в явном виде, лучше использовать цикл в foreach-стиле

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +243,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    val number = convert(n, base)
+    for (i in 0..number.size - 1) {
+        if (number[i] >= 10) result += ('a' - 10 + number[i]).toString()
+        else result += "${number[i]}"
```
Лучше через `toString()`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +261,26 @@ fun convertToString(n: Int, base: Int): String = TODO()
+            result *= n
+        }
+        return result
+    } else if (n != 0) return result + 1
```
Зачем здесь возвращать через `result + 1`, если можно сразу вернуть 1?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +261,26 @@ fun convertToString(n: Int, base: Int): String = TODO()
+        }
+        return result
+    } else if (n != 0) return result + 1
+    else return result
```
А здесь можно сразу вернуть 0

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +261,26 @@ fun convertToString(n: Int, base: Int): String = TODO()
+fun pow(base: Int, n: Int): Int {
+    var result = 0
+    if (base != 0) {
+        result++
```
Эту переменную можно просто завести здесь и проинициализировать единицей

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +261,26 @@ fun convertToString(n: Int, base: Int): String = TODO()
+fun decimal(digits: List<Int>, base: Int): Int {
+    var result = 0
+    for (i in 0..digits.size - 1) {
+        result += digits[i] * pow((digits.size - 1 - i), base)
```
Лишние скобки вокруг законченных выражений

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +291,13 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
-fun decimalFromString(str: String, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    var list = listOf<Int>()
+    for (i in 0..str.length - 1)
```
Цикл в foreach-стиле

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +291,13 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+    var list = listOf<Int>()
+    for (i in 0..str.length - 1)
+        if (str[i] in '0'..'9') list += (str[i] - 48).toInt()
+        else list += ((str[i]).toInt() - 87)
```
Лучше сделать эти преобразования через операции над `Char`, а не через `toInt()` и магические константы

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +316,88 @@ fun roman(n: Int): String = TODO()
+(char100 == 17) -> "семнадцать"
+(char100 == 18) -> "восемнадцать"
+(char100 == 19) -> "девятнадцать"
+else -> "десять"
```
Старый код нужно удалять, а не комментировать. Он все равно сохраняется в системе контроля версий, и вы сможете его вернуть при необходимости.

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +62,72 @@ fun main(args: Array<String>) {
+    var temp = ""
+    var temp2 = ""
+    for (part in parts) {
+        if (n == 3) n += 100
```
Если вам нужно сделать разные действия над разными элементами списка, зачем цикл???

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +62,72 @@ fun main(args: Array<String>) {
+    val parts = str.split(" ")
+    var result = ""
+    var f = 0
+    if (formatStringToDateStrToDigit(str)) {
```
Зачем делать два набора практически одинаковых действий в двух функциях, при этом никак не переиспользуя результаты, я не понимаю. Пожалуйста, поговорите с вашим преподавателем на занятиях о том, как следует переделать эти функции.

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -69,7 +136,82 @@ fun dateStrToDigit(str: String): String = TODO()
+    val parts = digital.split(".")
+    var result = ""
+    var f = 0
+    if (formatStringToDateDigitToStr(digital)) {
```
См. замечания выше и поговорите с преподавателем и про эти функции тоже

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +225,31 @@ fun dateDigitToStr(digital: String): String = TODO()
+        } else break
+    return if (n == phone.length) true
+    else false
+}
```
Вместо всей этой функции можно просто сделать `phone.any / phone.all` с правильным условием

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +225,31 @@ fun dateDigitToStr(digital: String): String = TODO()
+}
+
+fun flattenPhoneNumber(phone: String): String {
+    var f = 0
```
Какой смысл несет эта переменная?

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +225,31 @@ fun dateDigitToStr(digital: String): String = TODO()
+                f = 1
+            }
+        }
+        return result
```
Весь цикл выше можно заменить на `phone.map / phone.filter` (смотря что вы пытаетесь здесь сделать)

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +261,64 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    }
+    return if ((f == -1) || (numbers == 0)) false
+    else true
+}
```
Всю эту функцию можно заменить либо на проверку на основе регулярного выражения, либо (опять же), на `phone.all / phone.any`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +261,64 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+        else true
+    } else return if (str1.length > str2.length) true
+    else false
+}
```
Можно вообще обойтись без этой функции, так как строки и так можно сравнивать

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +261,64 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+            f = 0
+        }
+        return number.toInt()
+    } else return -1
```
Поговорите, пожалуйста, с преподавателем на занятиях о том, как следует правильно сделать эту функцию

ice-phoenix reviewed (changes requested):
Поговорите с преподавателем на занятиях о тех замечаниях, которые я указал. Плюс, обратите внимание, что не исправлены некоторые замечания из более ранних уроков.

Gukavina committed:
Исправления 2,3,4 уроков и "очень сложная" из 4ого урока

Gukavina committed:
В третьем сложная задача

ice-phoenix commented:
recheck all

ice-phoenix commented:
С учетом замечаний и статистики решения задач у вас "Удовлетворительно" автоматом.

Gukavina committed:
В третьем сложная задача

Gukavina committed:
Просто оставлю это здесь, ибо разбиралась с ним слишком много времени

Gukavina committed:
Просто оставлю это здесь, всё равно только сейчас увидела сообщение на почте, а урок уже сделала

ice-phoenix commented:
recheck all

ice-phoenix commented:
Даже с учетом дополнительно решенных задач все равно получается только "Удовлетворительно" автоматом.

Gukavina commented:
У меня есть ещё возможность исправить ошибки на этой неделе и получить оценку выше? Или уже не выйдет? Просто, как я понимаю, дело именно в неисправленных замечаниях, ибо по плану, который нам писали на лекции, у меня сделано в этом месяце на "4"

ice-phoenix commented:
На итоговое "Хорошо" необходимо было решить хотя бы одну задачку из 7.2. Кроме того, у нас есть некоторые вопросы к тому, как вы решили другие задачи. Попробуйте поговорить с преподавателем на упражнениях, но скорее всего ничего выше "Удовлетворительно" автоматом у вас не получится. Зато всегда можно прийти на экзамен и сдать на "Отлично"!

Gukavina committed:
7.2

Gukavina committed:
7.2
